### PR TITLE
New Button Mechanics

### PR DESCRIPTION
**Pull request by grant-m-s**
All the addition button mechanics have just built on top of the original buttonGroup mechanics. **All the code has also been networked.**

The current level has changed by having a rhythm button at the start (only req 1 player) and a spikes button trail mid-level (requires 3 players).

**Game Additions:**
Rhythm Buttons:
- Buttons that need to be pressed in a certain order.
- Sequence is determined by the order the positions are written.
- If buttons are not pressed in this order they reset so that the only button you are able to press is the first one
- If players are stepping on a button that shouldn't be stepped on then the buttons will be in a 'blocked' state where none of the buttons in the group will be able to be pressed
- Triggerable events (walls etc.) will only happen once the sequence has been completed.
- Multiple sets of buttons can be created where each player will have to complete their own sequence at the same time as another player. If any player presses the wrong button or misses a beat, all the buttons will become blocked and will reset once everyone has stepped off the buttons. This will force players to play to the beat and also makes sure their party is too.

Triggerable Spikes:
- Any button now has the ability to untrigger spikes in a level
- These can be the original style buttons or the new Rhythm style buttons

**File Changes:**
ComponentButtonGroup:
- Added `bool groupTriggerable` & `blocked` for use in rhythm buttons
- Added `int rhythmID` to keep track of which rhythm buttons belong in the same sequence
- Added `ivec2 prev` & `next` for sequencing
- Added `vector<ivec2> spikePositions`

ComponentSpikes:
- Keeps track of positions and whether triggered

Server Main:
- Added 2 helper functions for added buttons to level (`newButton` & `rhythmButton`). These are used to help the add the buttons to the level. They both make use of `tempo::createButtonGroup`.
- `newButton` reduces the need to write out parameters that are only required by rhythm buttons
- `rhythmButton` automatically assigns the `next`,`prev`,`rhythmID` parameters in a loop that goes through all the `buttonGroups` in the sequence.

SystemTrigger:
- if else statement to split normal buttons from rhythm buttons (normal buttons have `rhythmID = 0`)
- All rhythm button trigger logic
- Added two additional functions for resetting and locking button groups

SystemButtonRenderer:
- Updated for rhythm buttons
- Added new texture X for blocked buttons

Rendering Component & System for spikes

### CODE REVIEW

**Comment by grant-m-s** [comment]
The spike buttons can be tested just by opening another client as the ones in this level are just the normal ones.

I'll swap out the spike mesh soon.

I think the way its implemented is better as the whole idea is collaboration and rhythm if you let people move off the buttons then you've lost both of them in my opinion

**Comment by AnthonyWharton** [comment]
Cool, Cool, and fair enough. I just wonder if it could be annoying to be pinned to a button whilst your friends have to mess about. Will have to play test - fairly simple to implement I imagine :)

**Comment by grant-m-s** [comment]
I'm against leaving it a beat because once again the whole idea is rhythm. However having discussed with codrin, I will add arrow texture to top of buttons to show where the players need to go next

File: src/client/src/system/SystemButtonRenderer.cpp [hunk 1]
```diff
@@ -17,90 +17,259 @@ namespace client
...
+
+	int dx, dy;
+
+	dx = newPos.x - pos.x;
+
+	dy = newPos.y - pos.y;
+
+	int n = NAN; // for error checking
+
+	if (dx == 1) {
```
Comment (RcColes) [inline src/client/src/system/SystemButtonRenderer.cpp]: Can we not do something smarter with sin/cos/tan here

File: src/lib-tempo/include/tempo/network/ID.hpp [hunk 2]
```diff
@@ -55,16 +55,17 @@ enum ComponentID {
...
 	PLAYER_LOCAL,       // 10
 	PLAYER_REMOTE,      // 11
-	STAGE,              // 12
-	STAGE_POSITION,     // 13
-	STAGE_ROTATION,     // 14
-	STAGE_TRANSLATION,  // 15
-	TEAM,               // 16
-	TRANSFORM,          // 17
-	WEAPON,             // 18
+	SPIKES,				// 12
```
Comment (RcColes) [inline src/lib-tempo/include/tempo/network/ID.hpp]: Been indented with tabs instead of spaces for alignment

**Pull request merged**